### PR TITLE
Transfer method with tests

### PR DESCRIPTION
**Pull request by ghost**
Check Grid::transfer pls.

### CODE REVIEW

**Comment by Randl** [comment]
First, don't commit `cmake-build-debug ` (as well as any other build artifacts)

File: CMakeLists.txt [hunk 1]
```diff
@@ -3,4 +3,5 @@ project(LBM_CPP)
 
 set(CMAKE_CXX_STANDARD 11)
 
-add_executable(LBM_CPP main.cpp point.h constants.h)
\ No newline at end of file
+add_executable(LBM_CPP main.cpp point.h constants.h)
+add_subdirectory(tests)
```
Comment (Randl) [inline CMakeLists.txt]: Add newline at the end of file

File: main.cpp [hunk 2]
```diff
@@ -25,24 +25,25 @@ double*  vec_mul(double *a, double b) // vector * scalar
...
+  e[5].x = 1.;     e[5].y = 1.;
+  e[6].x = -1.;    e[6].y = 1.;
+  e[7].x = -1.;    e[7].y = -1.;
+  e[8].x = 1.;     e[8].y = -1.;
+  //std::vector<double> r = {1., 0.1, -0.1, 1.};
+  //p.macro();
+  //std::cout << p.tau << " " << p.rho << " " << p.T << " " << p.v.x << " "
+   //         << p.v.y << " ";
+  //std::cout << std::endl;
 }
```
Comment (Randl) [inline main.cpp]: Add newline at the end of file

File: point.h [hunk 3]
```diff
@@ -79,16 +82,29 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
       });
   xmin = indata[result.first - indata.begin()].first;
   xmax = indata[result.second - indata.begin()].first;
+  for (auto k : indata) {
+    k.first -= xmin;
+  }
+  xmin = 0;
+  xmax -= xmin;
```
Comment (Randl) [inline point.h]: This is meaningless, `xmin=0`

File: point.h [hunk 4]
```diff
@@ -79,16 +82,29 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
...
       [&](const std::pair<int, int> &a, const std::pair<int, int> &b) {
         return a.second < b.second;
       });
   ymin = indata[result.first - indata.begin()].second;
   ymax = indata[result.second - indata.begin()].second;
+  for (auto k : indata) {
+    k.second -= ymin;
+  }
+  ymin = 0;
+  ymax -= ymin;
```
Comment (Randl) [inline point.h]: This is meaningless, `ymin=0`

File: point.h [hunk 5]
```diff
@@ -106,17 +122,45 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
...
+          flag = true;
+        }
+        if (grid[x][y + e[k].y].bound) {
+          y_cord = -e[k].y;
+          flag = true;
+        }
+        if(!flag) {
+          x_cord = -e[k].x;
+          y_cord = -e[k].y;
+          flag = true;
```
Comment (Randl) [inline point.h]: why do you need it? `flag` will always be true.

File: point.h [hunk 6]
```diff
@@ -23,13 +22,17 @@ class Point {
...
+  void transfer(int, int);
   std::array<double, 5> macro_at(size_t, size_t); // TODO: result output
 private:
   std::vector<std::vector<Point>> grid;
+  FRIEND_TEST(GridConstructor__Test, MAX_MIN_Test);
+  FRIEND_TEST(GridBoundaries__Test, Boundaries_Test);
+  FRIEND_TEST(GridTransfer__Test, Transfer_Test);
 };
 
 void Point::col() {
```
Comment (Randl) [inline point.h]: maybe take out constants like `sc/c` and `v/c` in this function

File: point.h [hunk 7]
```diff
@@ -106,17 +122,45 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
...
+      for (size_t j = 0; j < Q; ++j) {
+        if (e[j].x == x_cord && e[j].y == y_cord) {
+          k_temp = j;
+          break;
+        }
       }
-      if (!grid[x + e[k].x][y].bound && grid[x][y + e[k].y].bound) {
-        e[k].x = -e[k].x;
+      if(flag) {
+        grid[x + e[k].x + e[k_temp].x][y + e[k].y + e[k_temp].y].f_temp[k_temp] =
```
Comment (Randl) [inline point.h]: you use `x + e[k].x` and `y + e[k].y` a lot. Define as a variable?

File: point.h [hunk 8]
```diff
@@ -125,14 +169,12 @@ void Grid::transfer(int x, int y) {
 void Grid::boundaries() {
   int count = 0;
   for (size_t i = 0; i < grid.size(); ++i) {
-    for (size_t j = 0; j < grid[i].size(); ++j) {
+    for (size_t j = 0; j < grid[0].size(); ++j) {
```
Comment (Randl) [inline point.h]: why replace `i` with `0`?

**Review by Randl — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (ghost) [reply CMakeLists.txt]: why should i do this?

Comment (Randl) [reply CMakeLists.txt]: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: point.h (hunk 5 above)
Comment (ghost) [reply point.h]: Here is an example of a situation when flag will be false
SharedScreenshot

Comment (Randl) [reply point.h]: You do `flag=true` in `if(flag==false)`. If this is intended, just put flag to be true on upper level

File: point.h (hunk 8 above)
Comment (ghost) [reply point.h]: it is the biggest one

Comment (Randl) [reply point.h]: If they are of different sizes you'll overflow. Else no reason to use 0

File: point.h [hunk 9]
```diff
@@ -97,14 +97,13 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
   for (auto k : indata) {
     k.second -= ymin;
   }
-  ymin = 0;
   ymax -= ymin;
```
Comment (EzAccount) [inline point.h]: declaration of ymin?

File: point.h [hunk 10]
```diff
@@ -106,17 +122,45 @@ Grid::Grid(std::vector<std::pair<int, int>> indata) {
...
-      if (!grid[x + e[k].x][y].bound && grid[x][y + e[k].y].bound) {
-        e[k].x = -e[k].x;
+      if(flag) {
+        grid[x + e[k].x + e[k_temp].x][y + e[k].y + e[k_temp].y].f_temp[k_temp] =
+            alpha * grid[x][y].f[k] +
+              (1 - alpha) *
+                (grid[x + e[k].x + e[k_temp].x][y + e[k].y + e[k_temp].y].f_eq[k_temp] -
+                   grid[x][y].f_eq[k_temp])/2;
+      } else {
+        grid[x + e[k].x][y + e[k].y].f_temp[k] =
```
Comment (EzAccount) [inline point.h]: Wrong. Planning to discuss over zoom Aug 8.

File: point.h [hunk 11]
```diff
@@ -124,22 +123,25 @@ void Grid::transfer(int x, int y) {
     int k_temp = 0;
     int x_cord = e[k].x, y_cord = e[k].y;
     int xOffset = x + e[k].x, yOffset = y + e[k].y;
-    bool flag =
-        false; // this flag is responsible for changing the direction of move
+    bool flag1 = false, flag2 = false;
```
Comment (Randl) [inline point.h]: Give meaningful names for variables

File: constants.h [hunk 12]
```diff
@@ -47,5 +47,6 @@ std::array<Vector<double>, Q> e;
 std::array<double, Q> w = {4. / 9,  1. / 9,  1. / 9,  1. / 9, 1. / 9,
                            1. / 36, 1. / 36, 1. / 36, 1. / 36};
 size_t x_size, y_size;
-
+size_t alpha = 0;
+size_t balance = 1;
```
Comment (Randl) [inline constants.h]: If those constants are not subject to change, make them `constexpr`. Else file naming is misleading

**Comment by Randl** [comment]
Note that "#X" is interpretated by git as "issue number X and "fixes #X" as "fix for issue number X". Keep it in mind when writing commit messages

**Pull request closed without merging**